newIDE/app/src/EventsSheet/InstructionEditor/NewInstructionEditorMenu.js

```diff
@@ -0,0 +1,351 @@
+// @flow
+import Popover from '@material-ui/core/Popover';
+
+import * as React from 'react';
+import {
+  type ResourceSource,
+  type ChooseResourceFunction,
+} from '../../ResourcesList/ResourceSource.flow';
+import { type ResourceExternalEditor } from '../../ResourcesList/ResourceExternalEditor.flow';
+import InstructionParametersEditor from './InstructionParametersEditor';
+import InstructionOrObjectSelector, {
+  type TabName,
+} from './InstructionOrObjectSelector';
+import {
+  createTree,
+  type InstructionOrExpressionTreeNode,
+} from './InstructionOrExpressionSelector/CreateTree';
+import {
+  type EnumeratedInstructionOrExpressionMetadata,
+  filterEnumeratedInstructionOrExpressionMetadataByScope,
+} from './InstructionOrExpressionSelector/EnumeratedInstructionOrExpressionMetadata.js';
+import InstructionOrExpressionSelector from './InstructionOrExpressionSelector';
+import {
+  enumerateObjectInstructions,
+  enumerateInstructions,
+  getObjectParameterIndex,
+} from './InstructionOrExpressionSelector/EnumerateInstructions';
+import { type EventsScope } from '../EventsScope.flow';
+import { SelectColumns } from '../../UI/Reponsive/SelectColumns';
+
+const gd = global.gd;
+
+const styles = {
+  fullHeightSelector: {
+    flex: 1,
+    display: 'flex',
+    flexDirection: 'column',
+    height: '400px',
+    width: '400px',
+  },
+};
+
+type StepName = 'object-or-free-instructions' | 'object-instructions';
+
+type Props = {|
+  project: gdProject,
+  scope: EventsScope,
+  globalObjectsContainer: gdObjectsContainer,
+  objectsContainer: gdObjectsContainer,
+  instruction: gdInstruction,
+  isCondition: boolean,
+  resourceSources: Array<ResourceSource>,
+  onChooseResource: ChooseResourceFunction,
+  resourceExternalEditors: Array<ResourceExternalEditor>,
+  style?: Object,
+  anchorEl?: any,
+  isNewInstruction: boolean,
+  onCancel: () => void,
+  onSubmit: () => void,
+  open: boolean,
+  openInstructionOrExpression: (
+    extension: gdPlatformExtension,
+    type: string
+  ) => void,
+|};
+type State = {|
+  step: StepName,
+  chosenObjectName: ?string,
+  chosenObjectInstructionsInfo: ?Array<EnumeratedInstructionOrExpressionMetadata>,
+  chosenObjectInstructionsInfoTree: ?InstructionOrExpressionTreeNode,
+  currentInstructionOrObjectSelectorTab: TabName,
+|};
+
+const findInstruction = (
+  list: Array<EnumeratedInstructionOrExpressionMetadata>,
+  instructionType: string
+): ?EnumeratedInstructionOrExpressionMetadata => {
+  return list.find(({ type }) => type === instructionType);
+};
+
+// TODO move to utils
```

Comment:
There is indeed some clever refactoring to do here - and maybe a bit of the logic inside the component methods too, to avoid duplication with the existing editor.
A bit of copy-pasta is not a big deal, but as possible would be good to reduce the copied lines, especially considered that there is also the old editor (but that will be removed sooner or later).

For now don't worry about this though! Let's get right the rest and I'm happy to help identifying what we could factor. In a sense, I prefer a bit of copy-pasta rather than too much refactoring that make the code looks like spagetthi code :)

Reply:
Was wondering where in utils to move this. I really dont like the copy pasta

Reply:
This should be a helper file in the same directory I think

Reply:
Utils.js ?